### ISSUE TITLE
Pulse_counter and Integration result in inf or -inf

### ISSUE THREAD

**Issue by FrankBakkerNl** [esphome/issues#1333]
**Operating environment/Installation (Hass.io/Docker/pip/etc.):**

ESP8266 D1 MINI and blitzwolf bw-shp6

ESP Home Version: 1.14.4

Pulsecounter / hlw8012 / integration

When configuring a pulse counter or the hlw8012 (which seems to use a pulsecounter internally) in combination with an integration sensor, the reported value sometimes reaches inf (infinate) or -inf). Once that happens it will never get back to a normal value as (inf + whatever stays inf). I experienced this first while using the pulse_counter with the integration sensor. Then later I had the same problem with the hlw8012.

**Description of problem:**

**Problem-relevant YAML-configuration entries:**
```yaml
  - platform: hlw8012
    sel_pin:
      number: GPIO12
      inverted: true
    cf_pin: GPIO05
    cf1_pin: GPIO14
    current_resistor: ${current_res}
    voltage_divider: ${voltage_div}
    current:
      name: '${device_name} Current'
      unit_of_measurement: 'A'
      accuracy_decimals: 3
      icon: mdi:flash-outline
    voltage:
      name: '${device_name} Voltage'
      unit_of_measurement: 'V'
      icon: mdi:flash-outline
    power:
      name: '${device_name} Power'
      id: power
      unit_of_measurement: 'W'
      icon: mdi:flash-outline
    change_mode_every: 4
    update_interval: 10s


  - platform: integration
    name: "${device_name} Total Energy "
    sensor: power
    time_unit: h
    unit_of_measurement: 'Wh'
    restore: true
    integration_method: left
```

**Logs (if applicable):**

```
[14:47:45][D][sensor:092]: 'esp_meter_2 Voltage': Sending state 0.00000 V with 1 decimals of accuracy    
[14:47:45][D][sensor:092]: 'esp_meter_2 Power': Sending state 0.00000 W with 1 decimals of accuracy      
[14:47:45][D][sensor:092]: 'esp_meter_2 Total Energy ': Sending state -inf Wh with 3 decimals of accuracy
```

**Additional information and things you've tried:**

AFAIK, the most common cause for a float to become inf is a division by zero. As the Inegration does not do any devisions that could be 0, I think the problem might be in the pulsecounter, specifically [line 157](https://github.com/esphome/esphome/blob/d4e76185bdb34f31a1be8efcbf04226b7a06596b/esphome/components/pulse_counter/pulse_counter_sensor.cpp#L157). Maybe the update_interval is not always initialized correcty?

**Comment by FrankBakkerNl** [comment]
As a workaround I added a filter to the sensor
```
      filters:
        # Prevent inf or -inf values
        - lambda: "return x == -(1.0 / 0.0) || x == (1.0 / 0.0) ? 0 : x;"
````
I cannot be sure this resolves it, but I have not had the problem since I added this

Is this maybe related to #1332 st7789v update interval is 0, causing home assistant api issues

**Comment by FrankBakkerNl** [comment]
As discussed with @OttoWinter in https://github.com/esphome/esphome/pull/1141 The problem is probably not in the pulse counter. I suspect it is caused by the 'restore: true' option of the integration sensor.
specifically 

https://github.com/esphome/esphome/blob/6405799cc2135b658da159b161ee4a50792b775a/esphome/components/integration/integration_sensor.cpp#L12
```
// integration_sensor.cpp
    this->rtc_ = global_preferences.make_preference<float>(this->get_object_id_hash());
    this->rtc_.load(&this->result_);

// integration_sensor.h
  double result_{0.0f};
```

I found the same problem with INF values when I crated a custom component which I first copy from the integration sensor and made some changes. I found that saving a value to the preferences and re-loading it sometimes resulted in INF. 
I suspect it is because the code uses `make_preference<float>`, while the field `result_` is of type `double` and this results in garbage being read.
In my custom component I have not had the problem after declaring the field as a float
